### PR TITLE
Avoid blocking when compiling Razor Pages

### PR DESCRIPTION
**Pull request by pranavkm**
Partial fix to https://github.com/aspnet/AspNetCore/issues/4880

We currently block as part of loading \ compiling a Razor Page and producing the cache that the action invoker uses to execute the page. This change defers the caching to a delegating invoker that executes asynchronously.

### CODE REVIEW

File: src/Mvc/src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/CreatePageCacheEntryActionInvoker.cs [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+        private readonly IModelBinderFactory _modelBinderFactory;
+        private readonly IFilterProvider[] _filterProviders;
+        private readonly IRazorPageFactoryProvider _razorPageFactoryProvider;
+        private readonly ParameterBinder _parameterBinder;
+        private readonly IModelMetadataProvider _modelMetadataProvider;
+        private readonly IPageActionInvokerFactory _pageActionInvokerFactory;
+        private readonly ConcurrentDictionary<ActionDescriptor, PageActionInvokerCacheEntry> _cache;
+
+        private readonly PageActionDescriptor _actionDescriptor;
+        private readonly ActionContext _actionContext;
```
Comment (rynowak) [inline src/Mvc/src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/CreatePageCacheEntryActionInvoker.cs]: This is going to be a **FAT** object that gets allocated each request. I do have a concern about the number of fields here.

File: src/Mvc/src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/CreatePageCacheEntryActionInvoker.cs [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+
+        public async Task InvokeAsync()
+        {
+            var compiledPageActionDescriptor = await _loader.LoadAsync(_actionDescriptor);
+            _actionContext.ActionDescriptor = compiledPageActionDescriptor;
+
+            var filterFactoryResult = FilterFactory.GetAllFilters(_filterProviders, _actionContext);
+            var filters = filterFactoryResult.Filters;
+            var cacheEntry = CreateCacheEntry(compiledPageActionDescriptor, filterFactoryResult.CacheableFilters);
+            cacheEntry = _cache.GetOrAdd(compiledPageActionDescriptor, cacheEntry);
```
Comment (rynowak) [inline src/Mvc/src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/CreatePageCacheEntryActionInvoker.cs]: Does someone else read this cache? It seems like we're doing a lot of work before accessing it.

File: src/Mvc/src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/CreatePageCacheEntryActionInvoker.cs [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+                    _parameterBinder,
+                    _modelMetadataProvider,
+                    _modelBinderFactory,
+                    actionDescriptor,
+                    actionDescriptor.HandlerMethods[i]);
+            }
+
+            return results;
+        }
+    }
```
Comment (rynowak) [inline src/Mvc/src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/CreatePageCacheEntryActionInvoker.cs]: Assuming everything else here is just the standard boilerplate.

File: src/Mvc/src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/IPageActionInvokerFactory.cs [hunk 4]
```diff
@@ -0,0 +1,16 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.AspNetCore.Mvc.Abstractions;
+using Microsoft.AspNetCore.Mvc.Filters;
+
+namespace Microsoft.AspNetCore.Mvc.RazorPages.Infrastructure
+{
+    internal interface IPageActionInvokerFactory
```
Comment (rynowak) [inline src/Mvc/src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/IPageActionInvokerFactory.cs]: Can we get rid of this? I really don't like how many layers this has.

File: src/Mvc/src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/CreatePageCacheEntryActionInvoker.cs [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+
+namespace Microsoft.AspNetCore.Mvc.RazorPages.Infrastructure
+{
+    /// <summary>
+    /// Executing a Razor Page requires ensuring it's compiled which is an asynchronous operation. 
+    /// The <see cref="IActionInvokerProvider"/> pipeline is synchronous and consequently we cannot compile a Razor Page
+    /// as part of executing it without blocking. The <see cref="PageActionInvokerProvider"/> instead delegates to this type
+    /// to compile the Razor Page and cache the information required for page invocation.
+    /// </summary>
+    internal class CreatePageCacheEntryActionInvoker : IActionInvoker
```
Comment (rynowak) [inline src/Mvc/src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/CreatePageCacheEntryActionInvoker.cs]: What's the benefit of having this as a separate class? What if the default invoker for pages just did this.

**Comment by pranavkm** [comment]
Superseded by https://github.com/aspnet/AspNetCore/pull/7938

**Pull request closed without merging**